### ISSUE TITLE
inaccessible to "lay developer"

### ISSUE THREAD

**Issue by timeglider** [defuse/php-encryption#313]
Right now this material is pretty hard to grok from the pov of a "multi-hats" developer who may not have the depth y'all have. I'd suggest providing some clear (copy-paste), simple examples of using the library to a.) establish and place keys on a server; b.) encrypt a string and c.) decrypt that string. You seem to be aiming to clear the air around encryption (hashing vs. encrypting, etc), but really, the material is impossible for the average developer to implement and then build upon. Write the code that would actually be used in a vanilla PHP installation, using proper namespacing, etc.  Right now, things are practically in "pseudo-code". I'd say that documentation, for stuff like this, is nearly as important as the code itself. If you want people to use the code, make it usable.

**Comment by defuse** [comment]
Thank you for opening this issue! This is exactly the kind of feedback we need. 

The challenge we face as developers of this library is that it's not application-specific. This library doesn't "know" which kind of application it's being used in, so users still face the challenge of *determining whether or not crypto is even useful to their application, and if it is, then how?* To get around this problem, in the [Tutorial](https://github.com/defuse/php-encryption/blob/master/docs/Tutorial.md), I tried to guess what the two most common use cases might be. And then I tried to explain in as detailed a way as possible how to implement those things with this library. I avoided making assumptions about what kind of database the app is using, how it's structured, etc. So the result is definitely pseudocode, but (I hoped) what was left as pseudocode would just be things developers are already familiar with.

I completely agree with you that the documentation is just as important as the code, so if you have any ideas for improvements, I'd be glad to hear them! More details about what you found confusing would be awesome, too.

Ultimately, what I think this means is that even with an API as "high level" as the one this library provides, implementing crypto where it's needed is still a hard problem. I think what us crypto-savvy devs need to do is start implementing those "common use cases" as libraries, so that *all* the details of the cryptography are hidden. The first step is to find out what those use cases are.
